src/visitor.rs

```diff
@@ -0,0 +1,205 @@
+    ///
+    /// The returned visitor is used to handle the bare item.
+    ///
+    /// Parsing will be terminated early if an error is returned.
+    fn item(&mut self) -> Result<impl ItemVisitor<'a>, Self::Error>;
+
+    /// Called after all inner-list items have been parsed.
+    ///
+    /// The returned visitor is used to handle the inner list's parameters.
+    /// Return [`Ignored`] to silently discard all parameters.
+    ///
+    /// Parsing will be terminated early if an error is returned.
+    fn finish(self) -> Result<impl ParameterVisitor<'a>, Self::Error>;
+}
+
+/// A visitor whose methods are called during entry parsing.
+///
+/// The lifetime `'a` is the lifetime of the input.
+pub trait EntryVisitor<'a>: ItemVisitor<'a> {
+    /// Called before an inner list has been parsed.
+    ///
+    /// The returned visitor is used to handle the inner list.
+    ///
+    /// Parsing will be terminated early if an error is returned.
+    fn inner_list(self) -> Result<impl InnerListVisitor<'a>, Self::Error>;
+}
+
+/// A visitor whose methods are called during dictionary parsing.
+///
+/// The lifetime `'a` is the lifetime of the input.
+///
+/// Use this trait with
+/// [`Parser::parse_dictionary_with_visitor`][crate::Parser::parse_dictionary_with_visitor].
+pub trait DictionaryVisitor<'a> {
+    type Error: Error;
+
+    /// Called after a dictionary key has been parsed.
+    ///
+    /// The returned visitor is used to handle the associated value.
+    ///
+    /// Parsing will be terminated early if an error is returned.
+    ///
+    /// Note: Per [RFC 8941], when duplicate dictionary keys are encountered in
+    /// the same scope, all but the last instance are ignored. Implementations
+    /// of this trait must respect that requirement in order to comply with the
+    /// specification. For example, if dictionary entries are stored in a map,
+    /// earlier values for a given dictionary key must be overwritten by later
+    /// ones.
+    ///
+    /// [RFC 8941]: <https://httpwg.org/specs/rfc8941.html#parse-dictionary>
+    fn entry(&mut self, key: &'a KeyRef) -> Result<impl EntryVisitor<'a>, Self::Error>;
+}
+
+/// A visitor whose methods are called during list parsing.
+///
+/// The lifetime `'a` is the lifetime of the input.
+///
+/// Use this trait with
+/// [`Parser::parse_list_with_visitor`][crate::Parser::parse_list_with_visitor].
+pub trait ListVisitor<'a> {
+    type Error: Error;
+
+    /// Called before a list entry has been parsed.
+    ///
+    /// The returned visitor is used to handle the entry.
+    ///
+    /// Parsing will be terminated early if an error is returned.
+    fn entry(&mut self) -> Result<impl EntryVisitor<'a>, Self::Error>;
+}
+
+/// A visitor that can be used to silently discard structured-field parts.
+///
+/// Note that the discarded parts are still validated during parsing: syntactic
+/// errors in the input still cause parsing to fail even when this type is used.
```

Comment:
One question I was thinking about was whether parsing should fail before we pass any output to the visitor.
For example, whether we should parse the dictionary once with Ignored, and if that succeeds we could parse it again with the real visitor.
I expect that's overkill, but we should add a note to the trait that consumers may want to do that if processing of entries/parameters/etc consumes computer resources.

Reply:
Good suggestion. Rather than suggesting re-parsing, though, I'm going to highlight the ability of the visitor to retain data internally so it can be acted on after parsing is complete.